pkg/policies/policies.go

```diff
@@ -79,29 +129,131 @@ func (pv *PolicyVerifier) Verify(ctx context.Context) ([]*engine.PolicyViolation
 		if err != nil {
 			return nil, fmt.Errorf("failed to verify policy: %w", err)
 		}
-		violations = append(violations, res...)
 
 		// 5. Store result in the attestation itself (for the renderer to include them in the predicate)
-		pv.state.Attestation.Policies = append(pv.state.Attestation.Policies, &v12.Policy{
+		result = append(result, &v12.Policy{
 			Name:       spec.Metadata.Name,
-			Attachment: policyAtt,
-			Body:       string(script.Source),
+			Body:       base64.StdEncoding.EncodeToString(script.Source),
 			Violations: policyViolationsToAttestationViolations(res),
 		})
 	}
 
-	return violations, nil
+	return result, nil
 }
 
-func (pv *PolicyVerifier) loadSpec(attachment *v1.PolicyAttachment) (*v1.Policy, error) {
+func engineViolationsToAPIViolations(input []*engine.PolicyViolation) []*v12.Policy_Violation {
+	res := make([]*v12.Policy_Violation, 0)
+	for _, v := range input {
+		res = append(res, &v12.Policy_Violation{
+			Subject: v.Subject,
+			Message: v.Violation,
+		})
+	}
+
+	return res
+}
+
+func getMaterialContent(material *v12.Attestation_Material, artifactPath string) ([]byte, error) {
+	var rawMaterial []byte
+	var err error
+
+	// nolint: gocritic
+	if material.InlineCas {
+		rawMaterial = material.GetArtifact().GetContent()
+	} else if artifactPath == "" {
+		return nil, errors.New("artifact path required")
+	} else {
+		// read content from local filesystem
+		rawMaterial, err = os.ReadFile(artifactPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read material content: %w", err)
+		}
+	}
+	// special case for ATTESTATION materials, the statement needs to be extracted from the dsse wrapper.
+	if material.MaterialType == v1.CraftingSchema_Material_ATTESTATION {
+		var envelope dsse.Envelope
+		if err := json.Unmarshal(rawMaterial, &envelope); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal attestation material: %w", err)
+		}
+
+		_, err = base64.StdEncoding.Decode(rawMaterial, []byte(envelope.Payload))
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode attestation material: %w", err)
+		}
+	}
+
+	return rawMaterial, nil
+}
+
+func (pv *PolicyVerifier) requiredPoliciesForMaterial(material *v12.Attestation_Material) ([]*v1.PolicyAttachment, error) {
```

Comment:
Could you add a comment on what this method does? I mean, the different logic paths it goes through.

In another patch, it would be great to test this method so it's used as documentation too.

Reply:
Yes, it's tested thoroughly in policies_test.go. I'll add the comment.